### PR TITLE
[DPE-1759] Juju 3

### PR DESCRIPTION
**Pull request by dragomirp**
## Issue
We should run integration tests against both juju 3 and juju 2

## Solution
* Add support for running integration test on both juju versions

### CODE REVIEW

File: tests/integration/test_db.py [hunk 1]
```diff
@@ -315,6 +315,7 @@ async def test_weebl_db(ops_test: OpsTest, charm: str) -> None:
         await ops_test.model.remove_application("weebl", block_until_done=True)
 
 
+@pytest.mark.notjuju3
 async def test_canonical_livepatch_onprem_bundle_db(ops_test: OpsTest) -> None:
```
Comment (dragomirp) [inline tests/integration/test_db.py:319]: Cannot deploy on juu 3

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -70,20 +70,26 @@ jobs:
...
           - password-rotation-integration
           - plugins-integration
           - tls-integration
-    name: ${{ matrix.tox-environments }}
+        agent-versions:
+          - "2.9.43"  # renovate: latest juju 2
+          - "3.1.5"  # renovate: latest juju 3
+        include:
+          - tox-environments: db-admin-relation-integration
+            agent-versions: "2.9.43"  # renovate: latest juju 2
```
Comment (dragomirp) [inline .github/workflows/ci.yaml:91]: Cannot deploy on juju 3

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -116,7 +129,7 @@ jobs:
             echo "mark_expression=not unstable" >> $GITHUB_OUTPUT
           fi
       - name: Run integration tests
-        run: tox run -e ${{ matrix.tox-environments }} -- -m '${{ steps.select-tests.outputs.mark_expression }}'
+        run: tox run -e ${{ matrix.tox-environments }}-${{ env.libjuju }} -- -m 'not not${{ env.libjuju }} and ${{ steps.select-tests.outputs.mark_expression }}'
```
Comment (dragomirp) [inline .github/workflows/ci.yaml:132]: Adding a mark to be able to skip tests based on Juju version.

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -804,8 +799,8 @@ async def run_command_on_unit(ops_test: OpsTest, unit_name: str, command: str) -
     Returns:
         the command output if it succeeds, otherwise raises an exception.
     """
-    complete_command = f"run --unit {unit_name} -- {command}"
```
Comment (dragomirp) [inline tests/integration/helpers.py:807]: Run seems to be used only for actions with Juju 3

File: tests/integration/helpers.py [hunk 5]
```diff
@@ -336,7 +337,7 @@ async def deploy_and_relate_bundle_with_postgresql(
         overlay: Optional overlay to be used when deploying the bundle.
     """
     # Deploy the bundle.
-    with tempfile.NamedTemporaryFile() as original:
+    with tempfile.NamedTemporaryFile(dir=os.getcwd()) as original:
```
Comment (dragomirp) [inline tests/integration/helpers.py:340]: Strictly confined juju3 snap has a separate tmp directory

**Review by marceloneppel — APPROVED** [review]
Looks great!

File: tests/integration/ha_tests/application-charm/requirements.txt
Comment (dragomirp) [inline tests/integration/ha_tests/application-charm/requirements.txt:1]: Stripping the hashes since otherwise we can't pack.

**Review by taurus-forever — APPROVED** [review]
Super. One question about Juju .43 VS .44...

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (dragomirp) [reply .github/workflows/ci.yaml:91]: This is the agent used for bootstraping. IIRC 2.9.44 got released while I was already working on this, but https://github.com/canonical/postgresql-operator/pull/190 will move this pin.

**Pull request merged**